Material Text Selection Toolbar improvements

justinmc opened the pull request:
## Description

This PR refactors material/text_selection.dart into more reasonable classes, and it makes public a top-level TextSelectionToolbar class and the default Material text selection menu button TextSelectionToolbarTextButton.

See the [text selection toolbar customization design doc](https://docs.google.com/document/d/1hyay1TyUZB7DXzWw45-ruGMZM1yCMWEnC3DzOJYemuc/edit?usp=sharing) for full details.

<img width="330" alt="Screen Shot 2020-11-04 at 3 01 32 PM" src="https://user-images.githubusercontent.com/389558/98177066-a8332900-1eae-11eb-9a1e-79ea109522da.png">

The PR stops short of making any breaking changes, so the developer experience is still not perfect, just much better than before.  A larger refactor of the text editing code is coming ([design doc](https://docs.google.com/document/d/1QaVIr1bbOWJGNyyNsB75sXgTDRMuxP268sLGTBNDut0/edit#heading=h.dy33bukwh2dy)), and we should be able to improve the experience further there.

### Example
I've published a working example of how to create a custom menu with the code in this branch [in a repo here](https://github.com/justinmc/flutter-text-selection-menu-examples/blob/master/lib/custom_menu_page.dart).

### Newly public classes

#### TextSelectionToolbar
A fully-functional Android-style text selection menu.  Built by [TextSelectionControls's buildToolbar](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/text_selection.dart#L742).

There is a bit of math to do in order to calculate the parameters needed (see where I do this in [buildToolbar in the example](https://github.com/justinmc/flutter-text-selection-menu-examples/blob/ea5a45c395b4df5834228021b82d2cdebf2a91d3/lib/custom_menu_page.dart#L42)).  

#### TextSelectionToolbarTextButton
A standard Android-style text selection menu button.  Used to create custom buttons and pass them into TextSelectionToolbar.

### Private classes
I cleaned up many of the private classes as well, to make sure that they have single responsibilities that make sense.  We may want to make some of these public in the future if users are making very custom menus and asking for these abilities.

#### _TextSelectionToolbarLayoutDelegate
Centers the toolbar at the given anchor and ensures it remains on screen.

#### _TextSelectionToolbarOverflowable
Maintains the state of whether or not the overflow menu is open, and renders the rest of the menu with this state.

#### _TextSelectionToolbarRightEdgeAlign
Just lines up the right edge of the menu when overflow is open (a quirk of how the overflow menu lays out on Android).

#### _TextSelectionToolbarItemsLayout
This does all of the hard overflow work.  Detects overflow and lays out the buttons in the regular or overflow menu.

#### _TextSelectionToolbarContainer
Just the appearance of the background of the menu (outline and elevation).

## Related Issues

https://github.com/flutter/flutter/issues/36985
https://github.com/flutter/flutter/issues/31412

This is not a full solution for those issues yet.  It's still fairly verbose to do something simple like add a custom menu button.  I think we'll need to make a breaking change to TextSelectionControls to get an acceptible solution for those.

## Tests

  * Directly test new class TextSelectionToolbar's ability to handle overflow.
  * Directly test new TextSeelectionToolbarTextButton's sizing logic.
  * Test a fully working SelectableText with a custom TextSelectionMenu.

## Breaking Change

Not a breaking change, but does expose some things that were previously private.

justinmc committed:
Abstract out simple stateless classes for buttons and toolbar

justinmc committed:
Abstract out a widget that builds the menu given some children, including handling overflow

justinmc committed:
Widget hierarchy and a description for reference

justinmc committed:
Buttons passed in top level, _MaterialTextSelectionToolbarNew eliminated, tests should be passing

justinmc committed:
A clean way to pass in buttons while handling paste hiding

justinmc committed:
WIP Public classes

justinmc committed:
WIP Renaming, documenting, reorganizing public classes

justinmc committed:
Merge branch 'master' into text-selection-menu-customization

justinmc committed:
Custom example working with minimal code duplication

justinmc committed:
Handle toolbar position in buttons

justinmc committed:
Naming

justinmc committed:
Correct children comparison.

skia-gold commented:
Gold has detected about 15 untriaged digest(s) on patchset 12.
View them at https://flutter-gold.skia.org/cl/github/69428

justinmc committed:
Analyzer fix and old TODO

skia-gold commented:
Gold has detected about 15 untriaged digest(s) on patchset 13.
View them at https://flutter-gold.skia.org/cl/github/69428

justinmc committed:
Fix reversed logic on children change

justinmc committed:
Fix logic for button position

skia-gold commented:
Gold has detected about 15 untriaged digest(s) on patchset 14.
View them at https://flutter-gold.skia.org/cl/github/69428

justinmc committed:
Fix error when no items for menu

skia-gold commented:
Gold has detected about 18 untriaged digest(s) on patchset 16.
View them at https://flutter-gold.skia.org/cl/github/69428

justinmc committed:
Fix incorrect padding for single item in menu

justinmc committed:
Merge branch 'master' into text-selection-menu-customization
I took only my changes and then fixed NNBD problems.

justinmc committed:
Link to issue for overflow slot

justinmc committed:
Better name: _TextSelectionToolbarContainer

justinmc committed:
Split button into its own file and test it

justinmc committed:
Move TextSelectionToolbar to its own file and test it

justinmc committed:
Add missing license headers

justinmc commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -9,115 +9,153 @@ import 'package:flutter/scheduler.dart';
+
+  /// Builder for material-style copy/paste text selection toolbar.
+  @override
+  Widget buildToolbar(
```
My approach for this PR was to not break the TextSelectionControls and buildToolbar interfaces.

The class that I'm making public is TextSelectionToolbar and not the _TextSelectionControlsToolbar class that I'm instantiating here because _TextSelectionControlsToolbar is designed for the buildToolbar API, which may be changed in the future.  I wrote TextSelectionToolbar's API to be as straightforward as possible without depending on the existing buildToolbar API.  In the future when we do a breaking refactor of TextSelectionControls and buildToolbar, TextSelectionToolbar should continue to be used without changes, and we can hopefully get rid of _TextSelectionControlsToolbar.

justinmc committed:
Code/docs cleanup

justinmc committed:
Test a fully custom menu in SelectableText

justinmc committed:
License fail

justinmc committed:
Merge branch 'master' into text-selection-menu-customization

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar_text_button.dart:
```diff
@@ -0,0 +1,119 @@
+}
+
+/// A button styled like a Material native Android text selection menu button.
+class TextSelectionToolbarTextButton extends StatelessWidget {
```
@HansMuller I'm curious about your opinion on me creating this button, since you just overhauled the button classes.

justinmc committed:
Easier way to get index

justinmc committed:
Use Padding rather than calculating constraints

justinmc committed:
Let overflow buttons take the color of the background

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+
+  // Return the value that centers width as closely as possible to position
+  // while fitting inside of min and max.
+  static double _centerOn(double position, double width, double min, double max) {
```
Thanks for pointing this out, I tried playing around with it and I found that my constraints were not including the padding.

After fixing that, yes, the constraints will limit the child so that it doesn't exceed `max-min`.  If the developer does something tricky to exceed the width constraints, this logic will position it at the furthest left possible and let it overflow on the right.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+    super.didUpdateWidget(oldWidget);
+    int childIndex = 0;
+    final bool childrenEqual =
+        widget.children.length == oldWidget.children.length
```
I didn't know that was built in, thanks!

justinmc committed:
Code review

justinmc committed:
Align menu to inner edge, not right edge always. Pass through text direction to overlay

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+  }
+}
+
+class _TextSelectionToolbarRightEdgeAlignRenderBox extends RenderProxyBox {
```
Thanks, I've now updated it to be trailing edge aligned too.  I used a Directionality widget inside the Overlay in order to pass through the text direction.

justinmc replied:
Looks like native iOS also makes the menu go away on rotation. I'll open an issue for that.

justinmc replied:
https://github.com/flutter/flutter/issues/72004

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+  }
+}
+
+class _TextSelectionToolbarItemsLayoutRenderBox extends RenderBox with ContainerRenderObjectMixin<RenderBox, ToolbarItemsParentData> {
```
Ah thanks, this isn't the first time I have messed that up.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+      }
+    });
+
+    // If the last child overflows, but only because of the width of the
```
### Last child width
No, the relative width shouldn't matter.  This is about the case where the overflow button causes the overflow, like:

|--------- width ---------|
[button 1] [button 2] [overflow button]

If we just get rid of the overflow button, then everything fits and we don't need an overflow button.

### Wrong Constraints
In response to another comment above (https://github.com/flutter/flutter/pull/69428#discussion_r538878026), I noticed that I should be using padding to set the constraints of the menu properly.  Does that cover your concern here?

### Consistent Size
Good catch, I've created an issue for this: https://github.com/flutter/flutter/issues/72023.  I spent a bunch of time trying to fix this right now, but I don't think I can align the text in a TextButton.  This problem existed before this PR, so I'll come back to it in that issue later.

LongCatIsLooong replied:
Sorry I wasn't clear. The overflow button is always the first child so in the case that it is extremely wide and takes up more horizontal space than button 1 and button 2 combined, we need to check more than the last child. 

i.e., 
children in layout order:
|--------- width ---------|
[wiiiiiiiiiiiiiiiiiiiiiiiiiide overflow button][button 1] [button 2]

justinmc replied:
Ah that makes sense, you're right.  I'm creating a separate issue for this as well since it's not new to this PR and should be pretty unlikely: https://github.com/flutter/flutter/issues/72075

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+          math.max(child.size.height, nextSize.height),
+        );
+      } else {
+        childParentData.offset = Offset(0.0, overflowHeight);
```
RE the first comment, see https://github.com/flutter/flutter/issues/72023, which I mentioned above.  You're right, it's incorrect and should be fixed.

For the second question, yes it's possible, but I believe I handled it consistently with native from when I originally wrote this.  The menu will position itself as high as possible and then grow downward from there:

| Closed | Open |
| -- | -- |
| <img width="479" alt="Screen Shot 2020-12-09 at 2 08 40 PM" src="https://user-images.githubusercontent.com/389558/101694554-22b01500-3a28-11eb-9740-1b66446ca98b.png"> | <img width="457" alt="Screen Shot 2020-12-09 at 2 08 46 PM" src="https://user-images.githubusercontent.com/389558/101694562-2479d880-3a28-11eb-85e7-cfd5085a88f6.png"> |

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+        );
+      } else {
+        childParentData.offset = Offset(0.0, overflowHeight);
+        overflowHeight += child.size.height;
```
Are you thinking that the order of the menu items in the overflow menu is reversed when it's above/below the anchor?  You're right that this code will not reverse the order.  I just tried to check this on native Android but I couldn't find a case where there was more than 1 button overflowing.  I can dig into it if that's what you mean.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar_text_button.dart:
```diff
@@ -0,0 +1,119 @@
+        shape: const RoundedRectangleBorder(),
+        minimumSize: const Size(kMinInteractiveDimension, kMinInteractiveDimension),
+        padding: EdgeInsets.only(
+          left: _getLeftPadding(_position),
```
It looks like Flutter doesn't currently reverse the order of the menu items for RTL languages (https://github.com/flutter/flutter/issues/59306).  I've added a note to the issue to also adjust this padding based on the text direction when that is implemented.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+        + _kToolbarScreenPadding;
+    final double availableHeight = anchorAbove.dy
+      - paddingTop;
+    final bool fitsAbove = _kToolbarHeight <= availableHeight;
```
See my earlier comment https://github.com/flutter/flutter/pull/69428#discussion_r539686089

justinmc committed:
Code review renaming and docs

justinmc committed:
Delay doing asMap until needed

justinmc committed:
Use padding instead of index/total for buttons

justinmc committed:
Fix toolbar test after addition of padding made it not fit

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar_text_button.dart:
```diff
@@ -0,0 +1,119 @@
+  ///
+  /// See also:
+  ///   [total], which is the total number of children in the toolbar.
+  final int index;
```
Great idea, thanks!  The index/total felt awkward to me.  I swapped them for a `padding` parameter and a static method that converts index and total to padding.

justinmc committed:
Docs accuracy improvement

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,721 @@
+    // saved width for the new set of children, and it prevents AnimatedSize
+    // from animating the size change.
+    _containerKey = UniqueKey();
+    // If the menu items change, make sure the overflow menu is closed. This
```
Good catch, I've reworded it.

So this comment in `_reset` is dealing with a situation where the children change such that they were overflowing (and the overflow menu was open) before the change, and afterwards the children all fit in the normal menu without overflow.

It's true that the logic in the code you linked would show all of the children, not an empty menu.  However, without resetting _overflowOpen to false, other things would be broken and in a half open / half closed state.  I think the menu items might render vertically for example.

Anyway, I've removed the logical error that says the overflow menu would be empty and replaced it with a more specific description of what's going on.

justinmc commented:
@LongCatIsLooong This is ready for re-review.

Thanks for all of the great feedback.  I've moved several things into separate issues, so double check that you're ok with those not being fixed here.  All of them existed in our toolbar before this PR, I've just moved the code and made it public now.

justinmc committed:
Correct invalid docs reference

LongCatIsLooong reviewed (approved):
Didn't realize some of the code were not new. Sorry for the off-topic comments. The new API looks good to me :+1:

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,742 @@
+  ///
+  /// Useful for customizing the high-level background of the toolbar. The given
+  /// child Widget will contain all of the [children].
+  final _ToolbarBuilder toolbarBuilder;
```
I guess we should, since users will be passing in their own toolbarBuilders.  I thought about making this more generic and maybe naming it ChildWidgetBuilder, but since it's possible that we'll need to add a new toolbar-specific parameter to this in the future, I'll keep the name as ToolbarBuilder.

justinmc commented on packages/flutter/lib/src/material/text_selection_toolbar.dart:
```diff
@@ -0,0 +1,742 @@
+  /// See also:
+  ///   * [TextSelectionToolbarTextButton], which builds a default Material-
+  ///     style text selection toolbar text button.
+  final List<Widget> children;
```
There is an assertion in this file that `children.length > 0`, and in text_selection.dart's buildToolbar method there is a check that won't even build TextSelectionToolbar if there are no children.

I'll add another assertion to the top level TextSelectionToolbar and also note this in the docs.

justinmc committed:
Code review fixes

justinmc committed:
Merge branch 'master' into text-selection-menu-customization

justinmc committed:
Dont run selection menu test on web because web uses the native menu

justinmc commented:
No worries!  I'm happy to have caught all of those bugs now.  I'll plan to merge this when the tests go green.
